File: PinpadMediaManager/src/PinpadMediaManager.Core/Models/QkDetectionRequest.cs
namespace PinpadMediaManager.Core.Models;

public static class QkDetectionRequest
{
    public static string BuildPayload(string? interfaces, string? transactionName, string? formattedAmount)
    {
        interfaces ??= "";
        transactionName ??= "";
        formattedAmount ??= "";
        if (interfaces.Length != 0 && (interfaces.Length != 3 || interfaces.Any(c => c != '0' && c != '1') || interfaces == "000"))
            throw new ArgumentException("Select at least one reader; the mask must contain three 0/1 digits in MSR, CHIP, CONTACTLESS order.");
        if (transactionName.Length > 64 || formattedAmount.Length > 48 || transactionName.Any(char.IsControl) || formattedAmount.Any(char.IsControl))
            throw new ArgumentException("Use up to 64 characters for the transaction name and 48 for the amount, without control characters.");
        if (formattedAmount.Length != 0) return $"{interfaces}\u001c{transactionName}\u001c{formattedAmount}";
        if (transactionName.Length != 0) return $"{interfaces}\u001c{transactionName}";
        return interfaces;
    }
}

File: PinpadApp/app/src/main/java/one/globalconnect/pinpad/protocol/QkDetectionRequest.kt
package one.globalconnect.pinpad.protocol

/** Optional FS-separated QK fields. Display text never changes transaction processing data. */
data class QkDetectionRequest(
    val interfaces: String = "111",
    val transactionName: String? = null,
    val formattedAmount: String? = null,
) {
    val swipe get() = interfaces[0] == '1'
    val chip get() = interfaces[1] == '1'
    val contactless get() = interfaces[2] == '1'

    companion object {
        fun parse(payload: String): QkDetectionRequest? {
            val fields = payload.split('\u001C')
            if (fields.size > 3) return null
            val mask = fields[0].ifEmpty { "111" }
            if (!mask.matches(Regex("[01]{3}")) || mask == "000") return null
            val name = fields.getOrNull(1)?.takeIf { it.isNotBlank() }
            val amount = fields.getOrNull(2)?.takeIf { it.isNotBlank() }
            if ((name?.length ?: 0) > 64 || (amount?.length ?: 0) > 48) return null
            if (listOfNotNull(name, amount).any { value -> value.any { it.isISOControl() } }) return null
            return QkDetectionRequest(mask, name, amount)
        }
    }
}
